src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -1048,35 +1048,67 @@ public long GetLastTombstoneEtag(DocumentsOperationContext context, string colle
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static Document TableValueToDocument(DocumentsOperationContext context, ref TableValueReader tvr, bool skipValidationInDebug = false)
+        private static Document TableValueToDocument(DocumentsOperationContext context, ref TableValueReader tvr, DocumentFields fields = DocumentFields.All, bool skipValidationInDebug = false)
         {
-            var document = ParseDocument(context, ref tvr);
+            var document = ParseDocument(context, ref tvr, fields);
 #if DEBUG
             if (skipValidationInDebug == false)
             {
                 Transaction.DebugDisposeReaderAfterTransaction(context.Transaction.InnerTransaction, document.Data);
                 DocumentPutAction.AssertMetadataWasFiltered(document.Data);
                 AttachmentsStorage.AssertAttachments(document.Data, document.Flags);
-            } 
+            }
 #endif
             return document;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static Document ParseDocument(JsonOperationContext context, ref TableValueReader tvr)
+        private static Document ParseDocument(JsonOperationContext context, ref TableValueReader tvr, DocumentFields fields)
         {
-            var result = new Document
+            if (fields == DocumentFields.All)
```

Comment:
This is going to be the most common action, right?
The problem is that now this method is pretty big, and full of conditional.
Let's move all the checks after this if to a dedicated method.

Reply:
Done